diff -bur metakit-2.4.9.6.orig/unix/configure metakit-2.4.9.6/unix/configure
--- metakit-2.4.9.6.orig/unix/configure	Thu Dec  7 11:40:25 2006
+++ metakit-2.4.9.6/unix/configure	Thu Dec  7 11:40:40 2006
@@ -3402,7 +3402,7 @@
     esac
     SHLIB_SUFFIX=".sl"
     ;;
-  cyg*)
+  cyg*|MINGW*)
     LIB_SUFFIX=".lib"
     SHLIB_SUFFIX=".dll"
     ;;
diff -bur metakit-2.4.9.6.orig/unix/configure.in metakit-2.4.9.6/unix/configure.in
--- metakit-2.4.9.6.orig/unix/configure.in	Thu Dec  7 11:40:25 2006
+++ metakit-2.4.9.6/unix/configure.in	Thu Dec  7 11:40:41 2006
@@ -164,7 +164,7 @@
     esac
     SHLIB_SUFFIX=".sl"
     ;;
-  cyg*)
+  cyg*|MINGW*)
     LIB_SUFFIX=".lib"
     SHLIB_SUFFIX=".dll"
     ;;
Only in metakit-2.4.9.6/unix: configure.in.orig
Only in metakit-2.4.9.6/unix: configure.orig
