--- ased3.0/ased.tcl.orig	Sat Nov 27 15:44:26 2004
+++ ased3.0/ased.tcl	Thu Nov  9 13:26:44 2006
@@ -166,7 +166,7 @@
     # normal start of ASED
     set libDir [file join $ASEDsRootDir lib]
     set auto_path [linsert $auto_path 0 "$libDir"]
-    package require -exact BWidget 1.7
+    package require -exact BWidget 1.8
     Bitmap::get cut
     # this used for loading Bitmap namespace into the interpetator
     namespace eval Bitmap {
--- ased3.0/create.tcl.orig	Sat Jun 19 12:24:46 2004
+++ ased3.0/create.tcl	Fri Mar 30 10:58:58 2007
@@ -416,9 +416,9 @@
             {separator}\
             [list command [tr "Goto Line ..."] {} [tr "Goto Line"] {} -command Editor::gotoLineDlg ]\
             {separator}\
-            [list command [tr "Search ..."] {} [tr "Search dialog"] {} -command Editor::search_dialog ]\
+            [list command [tr "Search ..."] {} [tr "Search dialog"] {Ctrl f} -command Editor::search_dialog ]\
             [list command [tr "Search in files ..."] {} [tr "Search in files"] {} -command Editor::findInFiles]\
-            [list command [tr "Replace ..."] {} [tr "Replace dialog"] {} -command Editor::replace_dialog ]\
+            [list command [tr "Replace ..."] {} [tr "Replace dialog"] {Ctrl h} -command Editor::replace_dialog ]\
             {separator}\
             [list command [tr "Undo"] {} [tr "Undo"] {CtrlAlt u} -command Editor::undo ]\
             [list command [tr "Redo"] {} [tr "Redo"] {} -command Editor::redo ]\
