--- configure.orig	Wed Dec 13 15:49:22 2006
+++ configure	Wed Dec 13 15:51:10 2006
@@ -656,7 +656,7 @@
     echo $ac_n "checking host type""... $ac_c" 1>&6
 echo "configure:658: checking host type" >&5
     case "`uname -s`" in
-	*win32* | *WIN32* | *CYGWIN_NT*|*CYGWIN_98*|*CYGWIN_95*)
+	*win32* | *WIN32* | *MINGW* | *CYGWIN_NT*|*CYGWIN_98*|*CYGWIN_95*)
 	    PLATFORM=WIN32
 	    cat >> confdefs.h <<\EOF
 #define WINDOWS 1
@@ -749,7 +749,7 @@
 	CC=gcc
     else
 	case "`uname -s`" in
-	    *win32* | *WIN32* | *CYGWIN_NT* | *CYGWIN_98* | *CYGWIN_95*)
+	    *win32* | *WIN32* | *MINGW* | *CYGWIN_NT* | *CYGWIN_98* | *CYGWIN_95*)
 		CC=cl
 	    ;;
 	    *)
@@ -1295,6 +1295,10 @@
 	CYGPATH="cygpath -w"
 	RELPATH=".. .. bin"
     ;;
+    *MINGW*)
+	CYGPATH=echo
+	RELPATH="."
+    ;;
     *)
 	CYGPATH=echo
 	RELPATH=".."
@@ -1511,7 +1515,7 @@
 #--------------------------------------------------------------------
 
 case "`uname -s`" in
-    *win32* | *WIN32* | *CYGWIN_NT*|*CYGWIN_98*|*CYGWIN_95*)
+    *win32* | *WIN32* | *MINGW* | *CYGWIN_NT*|*CYGWIN_98*|*CYGWIN_95*)
 	cat >> confdefs.h <<EOF
 #define BUILD_${PACKAGE} 1
 EOF
@@ -1553,7 +1557,7 @@
 
 
 	case "`uname -s`" in
-	    *win32* | *WIN32* | *CYGWIN_NT* | *CYGWIN_98* | *CYGWIN_95*)
+	    *win32* | *WIN32* | *MINGW* | *CYGWIN_NT* | *CYGWIN_98* | *CYGWIN_95*)
 		    echo "$ac_t""yes" 1>&6
 		;;
 	    *)
@@ -1681,7 +1685,7 @@
 
 
     case "`uname -s`" in
-	*win32* | *WIN32* | *CYGWIN_NT* | *CYGWIN_98* | *CYGWIN_95*)
+	*win32* | *WIN32* | *MINGW* | *CYGWIN_NT* | *CYGWIN_98* | *CYGWIN_95*)
 	    tcl_dbgx=d
 	;;
 	*)
@@ -1793,6 +1797,15 @@
 		eval "${PACKAGE}_LIB_FILE=${PACKAGE}${UNSHARED_LIB_SUFFIX}"
 	    fi
 	;;
+	*MINGW*)
+	    if test "${SHARED_BUILD}" = "1" ; then
+		SHLIB_LD_LIBS="${TCL_BIN_DIR}/${TCL_STUB_LIB_FILE} ${TCL_SHLIB_LD_LIBS} ${SSL_LIB_DIR}/libcrypto.dll.a ${SSL_LIB_DIR}/libssl.dll.a"
+		eval "${PACKAGE}_LIB_FILE=${PACKAGE}${SHARED_LIB_SUFFIX}"
+		RANLIB=:
+	    else
+		eval "${PACKAGE}_LIB_FILE=${PACKAGE}${UNSHARED_LIB_SUFFIX}"
+	    fi
+	;;
 	*)
 	    if test "${SHARED_BUILD}" = "1" ; then
 		SHLIB_LD_LIBS="${TCL_STUB_LIB_SPEC} -L${SSL_LIB_DIR} -lssl -L${SSL_LIB_DIR} -lcrypto ${GCCPATH} ${GCCLIB}"
@@ -1805,7 +1818,7 @@
     esac
 else
     case "`uname -s`" in
-        *win32* | *WIN32* | *CYGWIN_NT*|*CYGWIN_98*|*CYGWIN_95*)
+        *win32* | *WIN32* | *MINGW* | *CYGWIN_NT*|*CYGWIN_98*|*CYGWIN_95*)
 	    if test "${SHARED_BUILD}" = "1" ; then
 	        SHLIB_LD_LIBS="\"`${CYGPATH} ${TCL_BIN_DIR}/${TCL_STUB_LIB_FILE}`\" ${TCL_SHLIB_LD_LIBS} \"`${CYGPATH} ${SSL_LIB_DIR}/sslc32.lib`\""
 	        eval "${PACKAGE}_LIB_FILE=${PACKAGE}${SHARED_LIB_SUFFIX}"
