diff -Nbur mkZiplib1.0.orig/Makefile mkZiplib1.0/Makefile
--- mkZiplib1.0.orig/Makefile	Thu Jan  1 00:00:00 1970
+++ mkZiplib1.0/Makefile	Wed Nov 29 10:19:21 2006
@@ -0,0 +1,30 @@
+#!/bin/sh
+#
+# Mkziplib MinGW gcc Makefile 
+# Copyright (c) 2006 Martin Matuska
+#
+MKZIPLIB_SHORT=10
+CURDIR=$(shell pwd)
+PREFIX?=${CURDIR}/../../build
+ZLIB=$(shell which zlib1.dll)
+
+all: build
+
+build: mkZiplib${MKZIPLIB_SHORT}.dll 
+
+strip: build
+	strip mkZiplib${MKZIPLIB_SHORT}.dll
+
+mkZiplib${MKZIPLIB_SHORT}.dll: mkZiplib${MKZIPLIB_SHORT}.o zip.o unzip.o
+	gcc -shared -o mkZiplib${MKZIPLIB_SHORT}.dll mkZiplib${MKZIPLIB_SHORT}.o zip.o unzip.o ${PREFIX}/bin/tcl84.dll ${ZLIB}
+
+mkZiplib${MKZIPLIB_SHORT}.o:
+	gcc -I${PREFIX}/include -c -o mkZiplib${MKZIPLIB_SHORT}.o mkZiplib${MKZIPLIB_SHORT}.c
+unzip.o:
+	gcc -c -o unzip.o unzip.c
+
+zip.o:
+	gcc -c -o zip.o zip.c
+
+clean:
+	rm -f *.o *.dll
