diff -bur xotcl-1.5.3.orig/library/store/XOTclGdbm/Makefile.in xotcl-1.5.3/library/store/XOTclGdbm/Makefile.in
--- xotcl-1.5.3.orig/library/store/XOTclGdbm/Makefile.in	Wed Dec  6 13:20:42 2006
+++ xotcl-1.5.3/library/store/XOTclGdbm/Makefile.in	Wed Dec  6 13:26:26 2006
@@ -200,10 +200,12 @@
 install-libraries: libraries
 	@mkdir -p $(DESTDIR)$(includedir)
 	@echo "Installing header files in $(DESTDIR)$(includedir)"
-	@for i in $(PKG_HEADERS) ; do \
+	@if test -n "$(PKG_HEADERS)" ; then \
+	  for i in "$(PKG_HEADERS)" ; do \
 	    echo "Installing $(srcdir)/$$i" ; \
 	    $(INSTALL_DATA) $(srcdir)/$$i $(DESTDIR)$(includedir) ; \
-	done;
+	  done; \
+	fi
 
 #========================================================================
 # Install documentation.  Unix manpages should go in the $(mandir)
diff -bur xotcl-1.5.3.orig/library/store/XOTclSdbm/Makefile.in xotcl-1.5.3/library/store/XOTclSdbm/Makefile.in
--- xotcl-1.5.3.orig/library/store/XOTclSdbm/Makefile.in	Wed Dec  6 13:20:42 2006
+++ xotcl-1.5.3/library/store/XOTclSdbm/Makefile.in	Wed Dec  6 13:25:35 2006
@@ -200,10 +200,12 @@
 install-libraries: libraries
 	@mkdir -p $(DESTDIR)$(includedir)
 	@echo "Installing header files in $(DESTDIR)$(includedir)"
-	@for i in $(PKG_HEADERS) ; do \
+	@if test -n "$(PKG_HEADERS)" ; then \
+	  for i in "$(PKG_HEADERS)" ; do \
 	    echo "Installing $(srcdir)/$$i" ; \
 	    $(INSTALL_DATA) $(srcdir)/$$i $(DESTDIR)$(includedir) ; \
-	done;
+	  done; \
+	fi
 
 #========================================================================
 # Install documentation.  Unix manpages should go in the $(mandir)
--- xotcl-1.5.3.orig/library/xml/TclExpat-1.1/configure.in	Wed Dec  6 13:20:42 2006
+++ xotcl-1.5.3/library/xml/TclExpat-1.1/configure.in	Wed Dec  6 14:01:25 2006
@@ -90,7 +90,12 @@
 # and PKG_TCL_SOURCES.
 #-----------------------------------------------------------------------
 
-TEA_ADD_SOURCES([xmltok.c unixfilemap.c xmlrole.c xmlwf.c codepage.c xmlparse.c hashtable.c tclexpat.c])
+if test "${TEA_PLATFORM}" = "windows" ; then
+    TEA_ADD_SOURCES([xmltok.c win32filemap.c xmlrole.c xmlwf.c codepage.c xmlparse.c hashtable.c tclexpat.c])
+else
+    TEA_ADD_SOURCES([xmltok.c unixfilemap.c xmlrole.c xmlwf.c codepage.c xmlparse.c hashtable.c tclexpat.c])
+fi
+
 TEA_ADD_HEADERS([])
 TEA_ADD_INCLUDES([-I${with_xotcl}/generic ${XOTCL_BUILD_INCLUDE_SPEC}])
 TEA_ADD_LIBS([$XOTCL_BUILD_STUB_LIB_SPEC])
--- xotcl-1.5.3.orig/library/xml/TclExpat-1.1/Makefile.in	Wed Dec  6 13:20:42 2006
+++ xotcl-1.5.3/library/xml/TclExpat-1.1/Makefile.in	Wed Dec  6 14:06:04 2006
@@ -200,10 +200,12 @@
 install-libraries: libraries
 	@mkdir -p $(DESTDIR)$(includedir)
 	@echo "Installing header files in $(DESTDIR)$(includedir)"
-	@for i in $(PKG_HEADERS) ; do \
+	@if test -n "$(PKG_HEADERS)" ; then \
+	for i in "$(PKG_HEADERS)" ; do \
 	    echo "Installing $(srcdir)/$$i" ; \
 	    $(INSTALL_DATA) $(srcdir)/$$i $(DESTDIR)$(includedir) ; \
-	done;
+	done; \
+	fi
 
 #========================================================================
 # Install documentation.  Unix manpages should go in the $(mandir)
--- xotcl-1.5.3.orig/configure.in	Wed Dec  6 13:20:41 2006
+++ xotcl-1.5.3/configure.in	Wed Dec  6 14:21:25 2006
@@ -361,6 +361,8 @@
 # XOTcl specific configs
 #
 
+if test "${TEA_PLATFORM}" != "windows" ; then
+
 XOTCL_BUILD_LIB_SPEC="-L`pwd` -lxotcl${PACKAGE_VERSION}"
 XOTCL_LIB_SPEC="-L${pkglibdir} -lxotcl${PACKAGE_VERSION}"
 
@@ -377,6 +379,17 @@
 	XOTCL_STUB_LIB_SPEC="-L${pkglibdir} -lxotclstub${PACKAGE_VERSION}"
 	AC_DEFINE(COMPILE_XOTCL_STUBS)
 fi 
+
+else
+
+XOTCL_BUILD_LIB_SPEC="`pwd`/${PKG_LIB_FILE}"
+XOTCL_LIB_SPEC="${pkglibdir}/${PKG_LIB_FILE}"
+XOTCL_BUILD_STUB_LIB_PATH="`pwd`/${PKG_STUB_LIB_FILE}"
+XOTCL_STUB_LIB_PATH="${pkglibdir}/${PKG_STUB_LIB_FILE}"
+XOTCL_BUILD_STUB_LIB_SPEC="`pwd`/${PKG_STUB_LIB_FILE}"
+XOTCL_STUB_LIB_SPEC="${pkglibdir}/${PKG_STUB_LIB_FILE}"
+
+fi
 
 AC_SUBST(SHARED_LIB_SUFFIX)
 AC_SUBST(UNSHARED_LIB_SUFFIX)
--- xotcl-1.5.3.orig/Makefile.in	Thu Sep 28 09:19:12 2006
+++ xotcl-1.5.3/Makefile.in	Fri Apr  6 12:42:12 2007
@@ -75,7 +75,7 @@
 #========================================================================
 
 PKG_SOURCES	= @PKG_SOURCES@
-PKG_OBJECTS	= @PKG_OBJECTS@
+PKG_OBJECTS	= @PKG_OBJECTS@ xotclrc.o
 
 PKG_STUB_SOURCES = @PKG_STUB_SOURCES@
 PKG_STUB_OBJECTS = @PKG_STUB_OBJECTS@
@@ -193,6 +193,7 @@
 AR		= @AR@
 CFLAGS		= @CFLAGS@
 COMPILE		= $(CC) $(DEFS) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS)
+WINDRES		= windres
 
 #========================================================================
 # Start of user-definable TARGETS section
@@ -400,6 +401,13 @@
 
 .c.@OBJEXT@:
 	$(COMPILE) -c `@CYGPATH@ $<` -o $@
+
+xotclrc.o:
+	$(WINDRES) $(INCLUDES) -DXOTCL_MAJOR_VERSION=@XOTCL_MAJOR_VERSION@ \
+		-DXOTCL_COMMAVER=@XOTCL_MAJOR_VERSION@,@XOTCL_MINOR_VERSION@,$(subst .,,@XOTCL_RELEASE_LEVEL@),0 \
+		-DXOTCL_FILEVER=\\\"@XOTCL_MAJOR_VERSION@@XOTCL_MINOR_VERSION@$(subst .,,@XOTCL_RELEASE_LEVEL@)\\\" \
+		-DXOTCL_VERSION=\\\"@XOTCL_MAJOR_VERSION@.@XOTCL_MINOR_VERSION@@XOTCL_RELEASE_LEVEL@\\\" \
+		-i win/rc/xotcl.rc -o xotclrc.o
 
 #========================================================================
 # xotcl shells
--- xotcl-1.5.3.orig/win/rc/xotcl.rc	Sun May 23 22:50:39 2004
+++ xotcl-1.5.3/win/rc/xotcl.rc	Fri Apr  6 12:30:07 2007
@@ -21,8 +21,8 @@
 
 
 VS_VERSION_INFO VERSIONINFO
- FILEVERSION	XOTCL_MAJOR_VERSION ,XOTCL_MINOR_VERSION ,0 , 0
- PRODUCTVERSION	XOTCL_MAJOR_VERSION ,XOTCL_MINOR_VERSION ,0 , 0
+ FILEVERSION	XOTCL_COMMAVER
+ PRODUCTVERSION	XOTCL_COMMAVER
  FILEFLAGSMASK	0x3fL
 #if DEBUG == 1
  FILEFLAGS 	VS_FF_DEBUG
@@ -39,7 +39,7 @@
 		BEGIN
 			VALUE "FileDescription", "XOTcl language extension for Tcl\0"
 			VALUE "Authors", "Gustaf Neumann and Uwe Zdun\0"
-			VALUE "OriginalFilename", "xotcl" STRINGIFY(JOIN(XOTCL_MAJOR_VERSION,XOTCL_MINOR_VERSION)) SUFFIX ".dll\0"
+			VALUE "OriginalFilename", "xotcl" XOTCL_FILEVER ".dll\0"
 			VALUE "CompanyName", "Vienna University of Economics and Business Administration, University of Essen\0"
 			VALUE "FileVersion", XOTCL_VERSION
 			VALUE "LegalCopyright", "Copyright \251 1999-2001\0"
--- xotcl-1.5.3.orig/library/store/TclGdbmStorage.xotcl	Mon Apr 16 13:01:57 2007
+++ xotcl-1.5.3/store/TclGdbmStorage.xotcl	Mon Apr 16 13:02:07 2007
@@ -2,7 +2,7 @@
 
 package provide xotcl::store::tclgdbm 0.84
 
-package require xotcl::store::gdbm
+package require tgdbm
 package require xotcl::store
 package require XOTcl
 
