--- library/store/XOTclGdbm/xotclgdbm.c.orig	Tue Apr 10 12:27:31 2007
+++ library/store/XOTclGdbm/xotclgdbm.c	Tue Apr 10 12:27:52 2007
@@ -23,7 +23,7 @@
 } db_t;
 
 void
-gdbmFatalFunc(char* message) {
+XOTclGdbmFatalFunc(char* message) {
   fprintf(stderr, "GDBM FATAL:\n%s\n", message);
   exit(1);
 }
@@ -50,7 +50,7 @@
   block_size = 0;
   mode = 0644;
   
-  db->db = gdbm_open(TclObjStr(objv[1]), block_size, flags, mode, gdbmFatalFunc);
+  db->db = gdbm_open(TclObjStr(objv[1]), block_size, flags, mode, XOTclGdbmFatalFunc);
 
   if (db->db == NULL) {  
     ckfree ((char*) db);
